src/controllers/discussionController.ts

```diff
@@ -0,0 +1,266 @@
+import {
+  BadRequestError,
+  NotFoundError,
+  UnauthenticatedError,
+} from "../errors";
+import Joi from "joi";
+import mongoose from "mongoose";
+import { IUser } from "../models/User";
+
+const discussionJoiSchema = Joi.object({
+  title: Joi.string().required(),
+  book: Joi.string().required(),
+  content: Joi.string().required(),
+  date: Joi.date().iso().required(),
+  participants: Joi.array().items(Joi.string()).optional(),
+  meetingLink: Joi.string().uri().required(),
+  createdBy: Joi.string().required(),
+});
+
+interface IGetDiscussionsQuery {
+  search?: string;
+  sort?: "latest" | "oldest";
+}
+
+interface IJoinDiscussionBody {
+  user: mongoose.Schema.Types.ObjectId;
+}
+
+const getAllDiscussions = async (
+  req: Request<{}, {}, {}, IGetDiscussionsQuery & { bookId?: string }>,
+  res: Response,
+  next: NextFunction
+) => {
+  const { search, sort, bookId } = req.query;
+
+  let query: any = {};
+
+  if (bookId) {
+    query.book = bookId;
+  }
+
+  if (search) {
+    query.title = { $regex: new RegExp(search, "i") };
+  }
+
+  const sortOptions: any = {};
+  if (sort) {
+    if (sort === "latest") sortOptions.date = -1;
+    else if (sort === "oldest") sortOptions.date = 1;
+  } else {
+    sortOptions.date = -1;
+  }
+
+  try {
+    const discussions = await Discussion.find(query).sort(sortOptions);
+    res.status(StatusCodes.OK).json({ discussions, count: discussions.length });
+  } catch (error) {
+    return next(error);
+  }
+};
+
+const getDiscussion = async (
+  req: Request<{ discussionId: string }>,
+  res: Response,
+  next: NextFunction
+) => {
+  const { discussionId } = req.params;
+
+  try {
+    const discussion = await Discussion.findById(discussionId);
+    if (!discussion) {
+      return next(new NotFoundError(`No discussion with id ${discussionId}`));
+    }
+    res.status(StatusCodes.OK).json({ discussion });
+  } catch (error) {
+    return next(error);
+  }
+};
+
+const createDiscussion = async (
+  req: Request<{}, {}, IDiscussion>,
+  res: Response,
+  next: NextFunction
+) => {
+  const { error } = discussionJoiSchema.validate(req.body);
+  if (error) {
+    return next(new BadRequestError(error.details[0].message));
+  }
+
+  const user: IUser | undefined = req.user;
+
+  if (!user) {
+    return next(new UnauthenticatedError("User is not authenticated"));
+  }
+
+  try {
+    req.body.createdBy = user.userId;
+    const discussion = await Discussion.create(req.body);
+    res.status(StatusCodes.CREATED).json({ discussion });
+  } catch (error) {
+    return next(error);
+  }
+};
+
+const deleteDiscussion = async (
+  req: Request<{ discussionId: string }>,
+  res: Response,
+  next: NextFunction
+) => {
+  const { discussionId } = req.params;
+  const user: IUser | undefined = req.user;
+
+  if (!user) {
+    return next(new UnauthenticatedError("User is not authenticated"));
+  }
+
+   const discussion = await Discussion.findById(discussionId);
```

Comment:
you can also do sth like:
`const discussion = await Discussion.findOne({
  id: ObjectId(discussionId),
  createdBy: user.userId
});

if (!discussion) {
  return next(new NotFoundError(`No discussion with id ${discussionId} or user is not authorized to delete`));
}

try {
  await discussion.deleteOne();
  res.status(StatusCodes.OK).send();
} catch (error) {
  return next(error);
}`

Reply:
In this case, I don't like that we're not clear on whether the user is unauthorized or if there's no discussion with the id ${discussionId}. Can I leave it like this for now?